Add the `ecs_simplified` blog post (#1513)
iasql/iasql-engine: site/blog/ecs-simplified.mdx

- Deploy a load balancer as the point-of-contact for your app
- Create a target group for the load balancer and register the ECS tasks in it
- Add a new listener to your load balancer and connect it to the target group
- Create a security group, and you need to allow the port your app is listening on in that security group
- Attach the security group above to your load balancer
- Create a CloudWatch log group for your ECS task
- Create an ECS cluster, and definitely the task definition as well
- Oh, and create an ECR repository to push your images to be run on the container

I'm not going to continue this long list, since I've already got a headache. Doing those steps manually is going to give you a headache as well, so why bother doing all those steps yourself and risking different errors you might face when deploying your containerized app? You don't really have the time for the random IAM-related errors AWS is demanding you to resolve. Besides, you already have your codebase ready and the `Dockerfile` is there, so why not just run a simple command doing something that should be simply done?

> In this post, we're going to discover an IaSQL module that's built to make deploying to ECS, simplified. Most of the details for deploying a container to ECS are the same (load balancers, security groups, IAM roles, etc), and we have created the `ecs_simplified` module for you so that you can give it any Github repo with a `Dockerfile` and get your app deployed to ECS in the fastest time possible, with scalability available! All the needed resources are going to be created automatically in your AWS account, and you'll have the full access to the details while you're gaining the benefit of a higher-level simple deployment.

Reviewer: technically it is called `ecs_simplified` at the moment, but we should probably rename it?

## ** An Example of `aws_ecs_simplified` Module **

> Let's say we are going to deploy [this simple Express.js app](https://github.com/iasql/iasql-engine/tree/main/examples/ecs-fargate/prisma/app) to the ECS. It has a `Dockerfile` ready which `npm install`s the `express` package and then starts the Express server which listens on the port `8088`.

Reviewer 2: The URL changes to reference `v0.0.22` is to make sure this blog post always points at valid URLs.

- Now the ECS is waiting for an image to be pushed to your ECR repository to run it. You can get the URI for the ECR repository by running the `SELECT repository_uri FROM ecs_simplified WHERE app_name = 'simple-express'` query.  You *could* build your docker image locally and then follow [Steps 2 and 4 from this guide](https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html) to connect your local docker CLI to your ECR repository and push that docker image into your ECR repository, but we have a simpler solution next.
  - In the next step, we'll automatically build an image for the code in Github repo and then push it to this URI (all through SQL and using another high-level function named `ecr_build`).
```sql
SELECT ecr_build(
  'https://github.com/iasql/iasql-engine/', -- the Github repo URL
  (SELECT id
    FROM repository
    WHERE repository_name = 'simple-express-repository')::varchar(255), -- ECR repo for the image to be pushed
  './examples/ecs-fargate/prisma/app', -- the subdirectory in Github repo
  'main',  -- the Github branch or ref
  NULL -- Github personal access token - can be omitted if public repository
);
```

> - This command tells IaSQL to go clone `iasql-engine` repository, build an image on the subdirectory specified, and then push it to the ECR repository created earlier by `ecs_simplified` module. Running the above command will automatically create a CodeBuild project and the related roles, etc. Then it'll start a build, and after it's successful all the created resources are deleted to ensure there won't be any additional charges to your AWS account.

Reviewer 2: ```suggestion
- This command tells IaSQL to go clone `iasql-engine` repository, build an image on the subdirectory specified, and then push it to the ECR repository created earlier by the `aws_ecs_simplified` module. Running the above command will automatically create a CodeBuild project and the related roles, etc. Then it'll start a build, and after it's successful all the created resources are deleted to ensure there won't be any additional charges to your AWS account.
```

- To access your app on the cloud, get the load balancer address and use your browser to access the live version of it:
```sql
SELECT load_balancer_dns
FROM ecs_simplified
WHERE app_name = 'simple-express';
```

> Then you can check if the server is running on the `url:8088/health` address.

Reviewer 2: Might be slightly clearer as:

```suggestion
Then you can check if the server is running on the `<load_balancer_dns value>:8088/health` address.
```

## Low-level Access to Resources

> So the `aws_ecs_simplified` module simplifies things, right? But what if you still need the level of control you had when you were doing all the steps manually? The traditional PaaS trade-off is that you can't grow your app beyond limitations as you don't have access to all the small details. The IaSQL approach is not limited in that way.

Reviewer 2: ```suggestion
So the `aws_ecs_simplified` module simplifies things, right? But what if you still need the level of control you had when you were doing all the steps manually? The traditional PaaS trade-off is that you can't grow your app beyond the built-in limitations as you don't have access to all the small details. The IaSQL approach is not limited in that way.
```

Let's say you want your ECS container to be able to use the AWS CLI to provision an EC2 instance, and for that purpose its IAM role needs `AmazonEC2FullAccess` policy to work properly. `aws_ecs_simplified` does not have a column to configure such a thing, but that doesn't mean we're stuck.

The good news is that you still have the full control over all resources in the deepest details. Let's fix your app's IAM role access by attaching the needed policy to its IAM role:
```sql
UPDATE iam_role
SET attached_policies_arns = attached_policies_arns ||
                             'arn:aws:iam::aws:policy/AmazonEC2FullAccess' -- attached_policies_arns is of text[] type
WHERE role_name = 'simple-express-ecs-task-exec-role';

```

> You want additional rules for the container's security group? No problem! just write the SQL and execute it, and it will be applied to the cloud within seconds. You want 3 copies of your container to be kept running with a round-robin load balancing on them? It's already there, just do an `UPDATE ecs_simplified SET desired_count = 3 WHERE app_name = 'simple_express';` and it's there for you.

Reviewer 2: ```suggestion
You want additional rules for the container's security group? No problem! Just write the SQL and execute it, and it will be applied to the cloud within seconds. You want 3 copies of your container to be kept running with a round-robin load balancing on them? It's already there, just do an `UPDATE ecs_simplified SET desired_count = 3 WHERE app_name = 'simple_express';` and it's there for you.
```

Reviewer 2: Also, maybe this instead:

```suggestion
You want additional rules for the container's security group? No problem! Just write the SQL, execute it, and `SELECT iasql_apply()` it and it will be applied to the cloud within seconds. You want 3 copies of your container to be kept running with a round-robin load balancing on them? It's already there, just do an `UPDATE ecs_simplified SET desired_count = 3 WHERE app_name = 'simple_express';`, `SELECT iasql_apply()` it and it's there for you.
```

```

## ** An Example of `ecs_simplified` Module **
Let's say we are going to deploy [this simple Express.js app](https://github.com/iasql/iasql-engine/tree/main/examples/ecs-fargate/prisma/app) to the ECS. It has a `Dockerfile` ready which `npm install`s the `express` package and then starts the Express server which listens on the port `8088`.

Assuming that you have an IaSQL database, let's go and deploy the above app to our AWS account.
```sql title="Deploy a simple Express.js app from a Github repository to ECS" showLineNumbers TheButton="Deploy Now"
SELECT iasql_install('aws_ecs_simplified');
INSERT INTO ecs_simplified(app_name, app_port, image_tag, public_ip) VALUES(
  'simple-express',
  8088,
  'latest',
  true
);
SELECT iasql_apply();

SELECT ecr_build(
  'https://github.com/iasql/iasql-engine/', -- the Github repo URL
  (SELECT id
    FROM repository
    WHERE repository_name = 'simple-express-repository')::varchar(255), -- ECR repo for the image to be pushed
  './examples/ecs-fargate/prisma/app', -- the subdirectory in Github repo
  'main', -- the Github branch or ref

>   '' -- Github personal access token - can be omitted if public repository

Reviewer 2: Now that it's the last argument, can we write `NULL` instead of an empty string, or does that still fail?

Author: It works with `NULL`. I've changed all occurrences to use `NULL` argument.

- After running the `iasql_apply()` function to completion all the necessary resources are now created on the cloud, so their ARNs are populated. You can verify this by looking at different tables, eg. `iam_role`.
- To get your load balancer address, you can easily run `SELECT load_balancer_dns FROM ecs_simplified WHERE app_name = 'simple-express'` query and get the URL to access your app.
- Now the ECS is waiting for an image to be pushed to your ECR repository to run it. You can get the URI for the ECR repository by running the `SELECT repository_uri FROM ecs_simplified WHERE app_name = 'simple-express'` query.
- In the next step, we'll automatically build an image for the code in Github repo and then push it to this URI (all through SQL and using another high-level function named `ecr_build`).
```sql
SELECT ecr_build(
  'https://github.com/iasql/iasql-engine/', -- the Github repo URL
  (SELECT id
    FROM repository
    WHERE repository_name = 'simple-express-repository')::varchar(255), -- ECR repo for the image to be pushed
  './examples/ecs-fargate/prisma/app', -- the subdirectory in Github repo
  'main',  -- the Github branch or ref

>   '' -- Github personal access token - can be omitted if public repository

Reviewer 2: Same comment: can we make this `NULL` instead of an empty string? That would be more idiomatic in SQL, I think.

Author: It does work. I've changed to `NULL`.

```sql title="Deploy the folder 'examples/ecs-fargate/prisma/app' of the 'iasql-engine' repo to ECS on my AWS account"
-- Install the "high-level" aws_ecs_simplified IaSQL module
SELECT iasql_install('aws_ecs_simplified');
-- Create the load balancer, ECS service, task definition, etc. All ready afterwards for us to build and push to deploy the app.
INSERT INTO ecs_simplified (app_name, app_port, image_tag, public_ip) VALUES ('simple-express', 8088, 'latest', true);
SELECT iasql_apply();
-- Use the "high-level" ecr_build function to build the folder "examples/ecs-fargate/prisma/app" from "iasql-engine" Github repo and push it to ECR
SELECT ecr_build(
  'https://github.com/iasql/iasql-engine/',
  (SELECT id FROM repository WHERE repository_name = 'simple-express-repository')::varchar(255),
  './examples/ecs-fargate/prisma/app',
  'ghp_XXX', -- replace your github personal access token here
  'main'
);

> ```

Reviewer 3: since this piece of code is meant to be an example I would remove the comments asking actions like `-- replace your github personal access token here` and even maybe rename the app and use more placeholders. Taking a quick view I got confused between this code and the one in the next section and thought it was some kind of duplication.

```sql title="Create a security group named 'allow-all' attached to the VPC with ID 'vpc-f56cff91'. Then add a rule to it allowing all incoming traffic on port '80' from '0.0.0.0/0'"
-- Install the aws_security_group IaSQL module
SELECT iasql_install('aws_security_group');
-- "Create a security group named `allow-all` attached to the VPC with ID `vpc-f56cff91`"
INSERT INTO security_group (description, group_name, vpc_id)
VALUES ('allows all incoming traffic on port 80', 'allow-all', (SELECT id FROM vpc WHERE vpc_id = 'vpc-08c0a68025c2983a2'));
-- "Then add a rule to it allowing all incoming traffic on port `80` from `0.0.0.0/0`."
INSERT INTO security_group_rule (is_egress, ip_protocol, from_port, to_port, cidr_ipv4, description, security_group_id)
VALUES (false, 'tcp', 80, 80, '0.0.0.0/8', 'allow all to 80', (SELECT id FROM security_group WHERE group_name = 'allow-all'));

> ```

Reviewer 3: Also, maybe we can move the structure of this section to put the SQL examples closer to the point they're describing. Not a blocker, just a suggestion.

I mean, talk about high-level modules and show the example, then talk about the low-level module and show the example.

```sql
SELECT iasql_install('aws_ecs_simplified');
```

> - This command installs the `aws_ecs_simplified` high-level module. We– at IaSQL– have created that module to make it easy to deploy containerized apps to ECS. The code for it is [here](https://github.com/iasql/iasql-engine/blob/v0.0.22/src/modules/0.0.23/aws_ecs_simplified/sql/after_install.sql). But IaSQL is so flexible that anyone can create their own high-level (and of course, low-level) modules and add it to IaSQL.

Reviewer 2: This URL is broken. You need the `0.0.22` in both parts for it to work.

FROM ecs_simplified
WHERE app_name = 'simple-express';
```
Then you can check if the server is running on the `url:8088/health` address.

## Low-level Access to Resources

> So the `aws_ecs_simplified` module simplifies things, right? But what if you still want the level of control you had when you were doing all the steps manually? You can't grow your app beyond limitations if you don't have access to all the small details. Let's say you want your ECS container to be able to use the AWS CLI to provision an EC2 instance, and for that purpose its IAM role needs `AmazonEC2FullAccess` policy to work properly. `aws_ecs_simplifiy` is good, but wasn't it better if we could just manage the lower-level details as well?

Reviewer 2: Nit: want -> need. Hopefully people who *want* to just spend time fiddling with stuff are not using a tool to reduce the amount of stuff they fiddle with. ;)

Also restructuring a little bit that's a bit awkward-reading to me. Feel free to disregard that point, though.

```suggestion
So the `aws_ecs_simplified` module simplifies things, right? But what if you still need the level of control you had when you were doing all the steps manually? The traditional PaaS trade-off is that you can't grow your app beyond limitations as you don't have access to all the small details. The IaSQL approach is not limited in that way.

Let's say you want your ECS container to be able to use the AWS CLI to provision an EC2 instance, and for that purpose its IAM role needs `AmazonEC2FullAccess` policy to work properly. `aws_ecs_simplified` does not have a column to configure such a thing, but that doesn't mean we're stuck.
```
